compiler/AST/foralls.cpp

```diff
@@ -230,3 +235,495 @@ bool astUnderFI(const Expr* ast, ForallIntents* fi) {
+  // after detuple parIdx->parIdxCopy and after the DefExprs for ind vars.
+  AList& indvars = fs->inductionVariables();
+  int idx = indvars.length;
+  if (idx == 1) {
+    userLoopBody->insertAtHead("'move'(%S,%S)", toDefExpr(indvars.head)->sym,
+                               parIdxCopy);
+  } else {
+    for_alist_backward(def, indvars)
+      userLoopBody->insertAtHead("'move'(%S,%S(%S))", toDefExpr(def)->sym,
+                                 parIdxCopy, new_IntSymbol(idx--));
+  }
+  // move induction variables' DefExprs
+  for_alist_backward(def, indvars)
+    userLoopBody->insertAtHead(def->remove());
+}
+
+static void resolveParallelIteratorAndIdxVar(ForallStmt* pfs,
+                                             CallExpr* iterCall,
+                                             bool gotSA)
+{
+  // The par iterator probably has been extendLeader()-ed for forall intents.
+  FnSymbol* parIter = iterCall->resolvedFunction();
+  resolveFns(parIter);
+
+  // Set QualifiedType of the index variable.
+  QualifiedType iType = fsIterYieldType(parIter, iterCall);
+  VarSymbol* idxVar = parIdxVar(pfs);
+  idxVar->type = iType.type();
+  idxVar->qual = iType.getQual();
+}
+
+static Expr* rebuildIterableCall(ForallStmt* pfs,
+                                 CallExpr* iterCall,
+                                 Expr* origExprFlw)
+{
+  INT_ASSERT(iterCall == pfs->firstIteratedExpr()); // still here?
+
+  if (!pfs->zippered()) {
+    INT_ASSERT(!iterCall->next);
+    // no tuple building here
+    return origExprFlw;
+  }
+
+  int origLength = pfs->iteratedExpressions().length;
+  CallExpr* result = new CallExpr("_build_tuple", origExprFlw);
+  while (Expr* curr = iterCall->next)
+    result->insertAtTail(curr->remove());
+
+  // todo: remove the assert and origLength
+  INT_ASSERT(result->numActuals() == origLength);
+  return result;
+}
+
+static void buildLeaderLoopBody(ForallStmt* pfs, Expr* iterExpr) {
+  bool zippered = pfs->zippered();
+
+  // Set up the follower call etc.
+
+  // Extract leadIdxCopy and its type.
+  VarSymbol* leadIdxCopy     = parIdxCopyVar(pfs);
+
+  BlockStmt* resultBlock     = new BlockStmt();
+  // cf in build.cpp: new ForLoop(leadIdx, leadIter, NULL, zippered)
+  BlockStmt* leadForLoop     = new BlockStmt();
+
+  VarSymbol* iterRec         = newTemp("chpl__iterLF"); // serial iter, LF case
+
+  VarSymbol* followIdx       = newTemp("chpl__followIdx");
+  VarSymbol* followIter      = newTemp("chpl__followIter");
+  BlockStmt* followBlock     = NULL;
+
+  iterRec->addFlag(FLAG_EXPR_TEMP);
+  iterRec->addFlag(FLAG_CHPL__ITER);
+
+  followIdx->addFlag(FLAG_INDEX_OF_INTEREST);
+
+  resultBlock->insertAtTail(new DefExpr(iterRec));
+  resultBlock->insertAtTail(new CallExpr(PRIM_MOVE, iterRec, iterExpr));
+  Expr* toNormalize = resultBlock->body.tail;
+
+  // loop body to refer to followIdx instead of leadIdxCopy
+  for_SymbolSymExprs(se, leadIdxCopy)
+    if (se->parentExpr->parentExpr != pfs->loopBody()) // heuristic
```

Comment:
I've learned to dread comments like that. If it's a *heuristic*, what you're actually saying is that this code is a bug. Why did you put that comment? Why isn't it a complete solution? Should you be checking for arbitrary levels of nesting with parentExpr here, e.g.?